### DISCUSSION TITLE
Physics Petition

### DISCUSSION THREAD

**Discussion by Regrad** [godotengine/godot-proposals#6687] [General Discussions]
My game studio is forced to switch to Unreal Engine as Godot physics don't actually work in 3D. We really don't want to do this, so I'll try as a last resort to asking the developers to put Bullet's physics back into the engine until the engine's own physics are ready. Anyone who thinks the same - please vote like and write a comment.

**Comment by AThousandShips** [comment]
The claim that "don't actually work" is just not true, this kind of misrepresentation is not productive

**Reply by thehatless** [reply]
I can confirm that Godot 3d physics don't work in any non-toy situation.

**Reply by sammonius** [reply]
@thehatless Have you tried just adding more iterations in the settings? (Just out of curiosity, not being defensive)

**Reply by Regrad** [reply]
what iterations are you referring to? Solver iterations? No. What can it help with?

**Reply by sammonius** [reply]
Sorry about the late reply. Solver iterations prevent things from going through each other, and make a bunch of other things more precise too. Most practical physics situations are pretty complex so the best way to predict its impact is probably just to try increasing it and see what happens.

**Comment by Regrad** [comment]
A concave collision causes a catastrophic performance degradation. Convex collision causes the character to constantly get stuck. continuous_cd does not work. The option to automatic create a collision from many simple shapes does not work. Only collisions made from simple shapes, such as a cylinder or a cube, work fine. This is not enough to make full-fledged games in 3D. For understanding, here's the game I'm currently working on:
https://www.youtube.com/@Hellbringers/videos

**Comment by Calinou** [comment]
There are no plans to readd Bullet to core in Godot 4.x, but anyone is free to port it as an extension (similar to [godot-jolt](https://github.com/godot-jolt/godot-jolt)). See https://github.com/godotengine/godot-proposals/discussions/5161 for rationale (similar reasons apply to both Bullet and Jolt).

**Comment by Regrad** [comment]
I understand your position, but I also ask you to understand ours. The developers did not provide properly functioning physics in Godot 4. Bringing to mind the internal physics of the engine can take a significant amount of time. At the moment, using the engine in 3D in games is almost impossible.

**Comment by AThousandShips** [comment]
The core reason it has been difficult to solve the existing bugs in GodotPhysics is that there have been a lack of contributors able to solve these, I'm not sure how you imagine that is going to be any different with reimplementing Bullet

**Comment by Regrad** [comment]
It might be easier to port a normally working Bullet physics from Godot 3 than to create internal engine physics. The built-in physics of the engine is extremely far from working condition. I'm not sure that it will be possible to complete it even in a year. Porting Bullet would make it possible to create 3D games, since at the moment it is impossible to make a 3D game more complicated than Tetris.

**Reply by AThousandShips** [reply]
It probably won't, porting it from 3.x isn't just a case of a drop-in replacement, and would take time away from actually fixing the bugs, which is being worked on, see godotengine/godot#70522, and is in no way guaranteed to be free of bugs or issues adapting it to the framework of 4.x, it's entirely counterproductive to focus effort on a temporary stopgap measure

Also you seem to work with the impression that *no* games worth designing uses capsules and simple shapes for collision...

**Reply by Regrad** [reply]
Like I said, cubes and capsules can also be used, but this is not enough to make most 3D games harder than Tetris.
The physics of the engine is not ready for normal use in 3D, Bullet will not be ported. What are we to do then? Forget about Godot 4 in 3D? What's the point of releasing it with Vulcan if we can't use it properly? Just make renders and cutscenes?

**Comment by Regrad** [comment]
https://github.com/z80/godot4_bullet_gdextension

A good person did the Bullet physics transfer, maybe there is an opportunity to include this in the main assembly?

Forum topic:
https://godotforums.org/d/33803-bullet-physics-in-godot-4/13

**Reply by Zireael07** [reply]
> maybe there is an opportunity to include this in the main assembly?

Including it in Godot core would mean Godot maintainers would have to maintain it. This was exactly what they don't want to do, why Bullet was dropped for 4.

**Reply by Regrad** [reply]
Unfortunately, the author did not succeed in making this a module. It's better to spend a little time adjusting for new versions and remove it when they finish the new physics than just live without a properly working physics.

**Reply by InfiniteProductions** [reply]
> Unfortunately, the author did not succeed in making this a module. It's better to spend a little time adjusting for new versions and remove it when they finish the new physics than just live without a properly working physics.

Of course, but the issue here is people will most likely use Bullet instead and the devs would have a lot less chance to improve their own physics engine, with less issue report and less people willing to tweak and improve it.
By giving Godot users no other way, it's forcing evolution for the best, with some pain along the way.

**Reply by Regrad** [reply]
So far, it is obvious that Godot's physics is not yet ready for normal operation. Now the problems of internal physics are obvious and we can start by correcting this. It seems to me that we should bring it to normal operation and only then remove Bullet support. Moreover, a kind person has already ported it to version 4. Otherwise, who will test it if it is impossible to create complex 3D games on it? People just won't use Godot for 3D.

**Comment by azur-wolve** [comment]
Bullet had a lot of problems on 3.x
Also concave shapes always have been known to be problematic in general.

The best way in general is to use simple collision shapes: box, capsule, cylinder, etc.

Reading all that you have said, not sure if you even know enough about physics, as it sounds more speculative than nothing else.
Why you would need complex collision shapes for an indie shooter?

3D Physics aren't complete yet as was stated during 4.0 release. Need to wait a bit.

Other optional physics for Godot:
* Bullet
* Jolt
* Rapier

**Reply by ExiledSomnambulist** [reply]
You could always build more complicated collision shapes out of primitives.

**Comment by Shakai-Dev** [comment]
Jolt is being implemented, you can use it instead.
